src/platform/forms-system/src/js/utilities/data/reduceErrors.js

```diff
@@ -0,0 +1,156 @@
+ * @property {object} schema - ignored
+ * @property {string} stack - error message used in processing
+ * @property {string[]} __errors - may contain multiple error messages for one
+ *   element
+ *
+ * There are four types of hardcoded validation error messages:
+ *
+ * min/max: {
+    property: 'instance.someProperty',
+    message: 'does not meet minimum length of 1',
+    name: 'minItems',
+    argument: 1,
+    schema: {...},
+    stack: 'instance.someProperty does not meet minimum length of 1'
+  }
+  * required: {
+    property: 'instance',
+    message: 'requires property "someProperty"',
+    name: 'required',
+    argument: 'someProperty',
+    schema: {...},
+    stack: 'instance requires property "someProperty"'
+  }
+  * array: {
+    property: 'instance.newDisabilities[0]',
+    message: 'requires property "cause"',
+    name: 'required',
+    argument: 'cause',
+    schema: {...},
+    stack: 'instance.newDisabilities[0] requires property "cause"'
+  }
+  * non-empty "__errors" array: {
+    __errors: ['error message']
+  }
+*/
+/**
+ * @typedef Form~formErrors
+ * @type {object}
+ * @property {Form~rawErrors}
+ * @property {Form~errors}
+ */
+/**
+ * Process rawErrors from jsonschema validator into a more friendly list
+ * @param {Form~formErrors} errors
+ * @param {object[]} pageList - list of all form pages from `form.pages`
+ */
+export const reduceErrors = (errors, pageList) =>
+  errors.reduce((acc, error) => {
+    const findErrors = (key, err) => {
+      if (typeof err === 'object') {
+        if (err.message) {
+          const property = err.property.replace(/instance\.?/, '') || '';
+          const instance = property.replace(/(\[\d+\])/, '');
+          const name = err.name === 'required' ? err.argument : instance;
+          /*
+           * There may be multiple errors for the same property, e.g.
+           * `contestedIssues` in form 996. It can have a custom message
+           * "Please select a contested issue" and a minItem error
+           * "Does not meet minimum length of 1"; there's no need to confuse
+           * anyone and show both
+          */
+          if (!errorExists(acc, name)) {
+            const { chapterKey = '', pageKey = '' } = getPropertyInfo(
+              pageList,
+              err.argument,
+              err.property.startsWith('instance.') ? instance : '',
+            );
+            acc.push({
+              name,
+              // property may be `array[0]`; we need to extract out the `[0]`
+              index: property.match(/\[(\d+)\]/)?.[1] || null,
+              message: err.stack,
+              chapterKey,
+              pageKey,
+            });
+          }
+          return null;
+        }
+        if (err.__errors && err.__errors.length && !errorExists(acc, key)) {
+          const { chapterKey = '', pageKey = '' } = getPropertyInfo(
+            pageList,
+            key,
+          );
+          acc.push({
+            name: key,
```

Comment:
Why is `name` set to `key` here, but to `name` above? :confused:

Reply:
I moved the `__error` code to the top, because it's shorter and easier to read... anyway, when `__error` isn't an empty array, it's contained immediately within it's named object, inside it's wrapper, whereas the other type of errors provide the named object, which doesn't match the parameter passed into the `findErrors` function until it is at the same level as the property it's searching for... I hope that makes sense. It makes sense in my head, and I am still trying to remember all my reasonings from code I wrote many months ago.